### PR TITLE
feat(controller): Enable hot-standby. Closes #4077

### LINKED ISSUE #4077: Controller high-availability (HA) with hot-standby
# Summary

Determine what happen when we schedule 2 replicas and then determine how to prevent it.

# Use Cases

Related to #4076, #3751, #3648

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).
* [x] I've signed the CLA.
* [ ] ~~I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.~~
* [x] My builds are green. Try syncing with master if they are not. 

Closes #4077

image
image

### CODE REVIEW

**Comment by alexec** [comment]
@mmpei would you like to review this please?

**Comment by alexec** [comment]
@rhysemmas would you also like to review this. I'm aware this is for workflows, not rollouts, but if we get it right here, we can just migrate.

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -185,37 +187,66 @@ func (wfc *WorkflowController) Run(ctx context.Context, wfWorkers, podWorkers in
...
+			LockConfig: resourcelock.ResourceLockConfig{Identity: nodeID},
+		},
+		ReleaseOnCancel: true,
+		LeaseDuration:   15 * time.Second,
+		RenewDeadline:   10 * time.Second,
+		RetryPeriod:     2 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				logCtx.Info("started leading")
+				ctx, cancel = context.WithCancel(ctx)
```
Comment (alexec) [inline workflow/controller/controller.go:223]: we must cancel these when we are no longer the leader

**Comment by alexec** [comment]
Put back into draft pending controller ref analysis (might be a better option for HA).
#4318

**Comment by agilgur5** [comment]
Hot standby / leader-election was implemented in #4622

**Pull request closed without merging**